File: server/bluetooth_bridge.py
"""
FireBot Local Bluetooth Bridge Server
======================================

This Python Flask server acts as a bridge between:
- React Web App (via HTTP/WebSocket on localhost:5000)
- ESP32 FireBot (via Bluetooth Classic)

Communication Flow:
React App → HTTP/WebSocket → Python Server → Bluetooth → ESP32 → I²C → Arduino Uno

Requirements:
- Python 3.8+
- Flask
- Flask-CORS
- Flask-SocketIO
- PyBluez (for Bluetooth communication)

Install dependencies:
    pip install flask flask-cors flask-socketio pybluez

Usage:
    python bluetooth_bridge.py
"""

from flask import Flask, request, jsonify
from flask_cors import CORS
from flask_socketio import SocketIO, emit
import bluetooth
import threading
import time
import json

app = Flask(__name__)
CORS(app)  # Enable CORS for React app
socketio = SocketIO(app, cors_allowed_origins="*")

# ========================================
# Configuration
# ========================================
ESP32_NAME = "FireBot"  # Bluetooth device name
ESP32_ADDRESS = None    # Will be discovered
bt_socket = None
is_connected = False
sensor_data = {
    "fireDetected": False,
    "pumpStatus": False,
    "connected": False
}

# ========================================
# Bluetooth Functions
# ========================================

def discover_esp32():
    """Discover ESP32 Bluetooth device"""
    global ESP32_ADDRESS
    
    print("🔍 Searching for ESP32 FireBot...")
    nearby_devices = bluetooth.discover_devices(duration=8, lookup_names=True)
    
    for addr, name in nearby_devices:
        if ESP32_NAME in name:
            ESP32_ADDRESS = addr
            print(f"✅ Found {name} at {addr}")
            return True
    
    print(f"❌ {ESP32_NAME} not found")
    return False

def connect_bluetooth():
    """Connect to ESP32 via Bluetooth Classic"""
    global bt_socket, is_connected, ESP32_ADDRESS
    
    try:
        if ESP32_ADDRESS is None:
            if not discover_esp32():
                return False
        
        print(f"📡 Connecting to {ESP32_ADDRESS}...")
        bt_socket = bluetooth.BluetoothSocket(bluetooth.RFCOMM)
        bt_socket.connect((ESP32_ADDRESS, 1))  # Port 1 for RFCOMM
        is_connected = True
        sensor_data["connected"] = True
        
        print("✅ Connected to ESP32 FireBot")
        
        # Start sensor data listener thread
        listener_thread = threading.Thread(target=sensor_listener, daemon=True)
        listener_thread.start()
        
        # Emit connection status via WebSocket
        socketio.emit('connection_status', {'connected': True})
        
        return True
        
    except Exception as e:
        print(f"❌ Bluetooth connection failed: {e}")
        is_connected = False
        sensor_data["connected"] = False
        return False

def disconnect_bluetooth():
    """Disconnect from ESP32"""
    global bt_socket, is_connected
    
    if bt_socket:
        try:
            bt_socket.close()
            print("🔌 Disconnected from ESP32")
        except:
            pass
    
    bt_socket = None
    is_connected = False
    sensor_data["connected"] = False
    socketio.emit('connection_status', {'connected': False})

def send_command(command):
    """Send command to ESP32 via Bluetooth"""
    global bt_socket, is_connected
    
    if not is_connected or bt_socket is None:
        return False
    
    try:
        # Send command as string
        bt_socket.send(command.encode('utf-8'))
        print(f"📤 Sent command: {command}")
        return True
    except Exception as e:
        print(f"❌ Failed to send command: {e}")
        is_connected = False
        sensor_data["connected"] = False
        return False

def sensor_listener():
    """Listen for sensor data from ESP32"""
    global bt_socket, is_connected, sensor_data
    
    while is_connected and bt_socket:
        try:
            # Receive data from ESP32
            data = bt_socket.recv(1024).decode('utf-8')
            
            if data:
                # Parse sensor data (format: "FIRE:0,DIST:0,PUMP:1")
                parts = data.strip().split(',')
                for part in parts:
                    if ':' in part:
                        key, value = part.split(':')
                        if key == 'FIRE':
                            sensor_data["fireDetected"] = (value == '1')
                        elif key == 'PUMP':
                            sensor_data["pumpStatus"] = (value == '1')
                
                # Emit sensor data via WebSocket
                socketio.emit('sensor_data', sensor_data)
                print(f"📊 Sensor data: {sensor_data}")
                
        except Exception as e:
            if is_connected:
                print(f"❌ Sensor listener error: {e}")
                is_connected = False
                sensor_data["connected"] = False
                socketio.emit('connection_status', {'connected': False})
            break
        
        time.sleep(0.1)  # Small delay

# ========================================
# REST API Endpoints
# ========================================

@app.route('/api/connect', methods=['POST'])
def api_connect():
    """Connect to ESP32 Bluetooth"""
    success = connect_bluetooth()
    return jsonify({
        'success': success,
        'connected': is_connected,
        'message': 'Connected to FireBot' if success else 'Connection failed'
    })

@app.route('/api/disconnect', methods=['POST'])
def api_disconnect():
    """Disconnect from ESP32"""
    disconnect_bluetooth()
    return jsonify({
        'success': True,
        'connected': False,
        'message': 'Disconnected from FireBot'
    })

@app.route('/api/send-command', methods=['POST'])
def api_send_command():
    """Send command to ESP32"""
    data = request.get_json()
    command = data.get('command', '')
    
    if not command:
        return jsonify({'success': False, 'message': 'No command provided'}), 400
    
    if not is_connected:
        return jsonify({'success': False, 'message': 'Not connected to FireBot'}), 503
    
    success = send_command(command)
    
    return jsonify({
        'success': success,
        'message': f'Command {command} sent' if success else 'Failed to send command'
    })

@app.route('/api/status', methods=['GET'])
def api_status():
    """Get current robot status"""
    return jsonify({
        'connected': is_connected,
        'sensorData': sensor_data
    })

@app.route('/api/health', methods=['GET'])
def api_health():
    """Health check endpoint"""
    return jsonify({
        'status': 'running',
        'server': 'FireBot Bluetooth Bridge',
        'version': '1.0.0'
    })

# ========================================
# WebSocket Events
# ========================================

@socketio.on('connect')
def handle_ws_connect():
    """WebSocket connection established"""
    print("🔌 WebSocket client connected")
    emit('connection_status', {'connected': is_connected})
    emit('sensor_data', sensor_data)

@socketio.on('disconnect')
def handle_ws_disconnect():
    """WebSocket client disconnected"""
    print("🔌 WebSocket client disconnected")

@socketio.on('send_command')
def handle_ws_command(data):
    """Handle command via WebSocket"""
    command = data.get('command', '')
    success = send_command(command)
    emit('command_response', {'success': success, 'command': command})

# ========================================
# Main
# ========================================

if __name__ == '__main__':
    print("\n" + "="*50)
    print("🔥 FireBot Local Bluetooth Bridge Server")
    print("="*50)
    print("Server running on: http://localhost:5000")
    print("WebSocket available on: ws://localhost:5000")
    print("\nEndpoints:")
    print("  POST /api/connect       - Connect to ESP32")
    print("  POST /api/disconnect    - Disconnect from ESP32")
    print("  POST /api/send-command  - Send command to robot")
    print("  GET  /api/status        - Get robot status")
    print("  GET  /api/health        - Server health check")
    print("\nPress Ctrl+C to stop")
    print("="*50 + "\n")
    
    # Run Flask-SocketIO server
    socketio.run(app, host='0.0.0.0', port=5000, debug=True, use_reloader=False)

File: server/bluetooth_bridge.js
/**
 * FireBot Local Bluetooth Bridge Server (Node.js)
 * =================================================
 * 
 * This Node.js Express server acts as a bridge between:
 * - React Web App (via HTTP/WebSocket on localhost:5000)
 * - ESP32 FireBot (via Bluetooth Serial)
 * 
 * Communication Flow:
 * React App → HTTP/WebSocket → Node Server → Bluetooth → ESP32 → I²C → Arduino Uno
 * 
 * Requirements:
 * - Node.js 16+
 * - Express
 * - Socket.IO
 * - serialport (for Bluetooth Serial communication)
 * 
 * Install dependencies:
 *     npm install express socket.io cors serialport
 * 
 * Usage:
 *     node bluetooth_bridge.js
 */

const express = require('express');
const http = require('http');
const socketIO = require('socket.io');
const cors = require('cors');
const { SerialPort } = require('serialport');
const { ReadlineParser } = require('@serialport/parser-readline');

const app = express();
const server = http.createServer(app);
const io = socketIO(server, {
  cors: {
    origin: "*",
    methods: ["GET", "POST"]
  }
});

// Middleware
app.use(cors());
app.use(express.json());

// ========================================
// Configuration
// ========================================
const PORT = 5000;
const BLUETOOTH_PORT = 'COM3'; // ⚠️ CHANGE THIS to your ESP32 COM port (check Device Manager)
const BAUD_RATE = 115200;

let serialPort = null;
let parser = null;
let isConnected = false;
let sensorData = {
  fireDetected: false,
  pumpStatus: false,
  connected: false
};

// ========================================
// Bluetooth Serial Functions
// ========================================

/**
 * Connect to ESP32 via Bluetooth Serial
 */
function connectBluetooth() {
  return new Promise((resolve, reject) => {
    try {
      console.log(`📡 Connecting to ESP32 on ${BLUETOOTH_PORT}...`);
      
      serialPort = new SerialPort({
        path: BLUETOOTH_PORT,
        baudRate: BAUD_RATE
      });

      parser = serialPort.pipe(new ReadlineParser({ delimiter: '\n' }));

      serialPort.on('open', () => {
        isConnected = true;
        sensorData.connected = true;
        console.log('✅ Connected to ESP32 FireBot');
        
        // Emit connection status to all WebSocket clients
        io.emit('connection_status', { connected: true });
        
        resolve(true);
      });

      serialPort.on('error', (err) => {
        console.error('❌ Serial port error:', err.message);
        isConnected = false;
        sensorData.connected = false;
        reject(err);
      });

      serialPort.on('close', () => {
        console.log('🔌 Disconnected from ESP32');
        isConnected = false;
        sensorData.connected = false;
        io.emit('connection_status', { connected: false });
      });

      // Listen for sensor data
      parser.on('data', (data) => {
        handleSensorData(data);
      });

    } catch (error) {
      console.error('❌ Bluetooth connection failed:', error);
      reject(error);
    }
  });
}

/**
 * Disconnect from ESP32
 */
function disconnectBluetooth() {
  return new Promise((resolve) => {
    if (serialPort && serialPort.isOpen) {
      serialPort.close((err) => {
        if (err) {
          console.error('Error closing port:', err);
        }
        isConnected = false;
        sensorData.connected = false;
        io.emit('connection_status', { connected: false });
        resolve();
      });
    } else {
      isConnected = false;
      sensorData.connected = false;
      resolve();
    }
  });
}

/**
 * Send command to ESP32
 */
function sendCommand(command) {
  return new Promise((resolve, reject) => {
    if (!isConnected || !serialPort || !serialPort.isOpen) {
      reject(new Error('Not connected to ESP32'));
      return;
    }

    serialPort.write(command + '\n', (err) => {
      if (err) {
        console.error('❌ Failed to send command:', err);
        reject(err);
      } else {
        console.log(`📤 Sent command: ${command}`);
        resolve(true);
      }
    });
  });
}

/**
 * Handle incoming sensor data from ESP32
 */
function handleSensorData(data) {
  try {
    // Parse sensor data (format: "FIRE:0,DIST:0,PUMP:1")
    const parts = data.trim().split(',');
    
    parts.forEach(part => {
      if (part.includes(':')) {
        const [key, value] = part.split(':');
        
        if (key === 'FIRE') {
          sensorData.fireDetected = (value === '1');
        } else if (key === 'PUMP') {
          sensorData.pumpStatus = (value === '1');
        }
      }
    });

    // Emit sensor data to all WebSocket clients
    io.emit('sensor_data', sensorData);
    console.log('📊 Sensor data:', sensorData);

  } catch (error) {
    console.error('Error parsing sensor data:', error);
  }
}

// ========================================
// REST API Endpoints
// ========================================

/**
 * POST /api/connect - Connect to ESP32 Bluetooth
 */
app.post('/api/connect', async (req, res) => {
  try {
    await connectBluetooth();
    res.json({
      success: true,
      connected: true,
      message: 'Connected to FireBot'
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      connected: false,
      message: 'Connection failed: ' + error.message
    });
  }
});

/**
 * POST /api/disconnect - Disconnect from ESP32
 */
app.post('/api/disconnect', async (req, res) => {
  try {
    await disconnectBluetooth();
    res.json({
      success: true,
      connected: false,
      message: 'Disconnected from FireBot'
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Disconnect failed: ' + error.message
    });
  }
});

/**
 * POST /api/send-command - Send command to ESP32
 */
app.post('/api/send-command', async (req, res) => {
  const { command } = req.body;

  if (!command) {
    return res.status(400).json({
      success: false,
      message: 'No command provided'
    });
  }

  if (!isConnected) {
    return res.status(503).json({
      success: false,
      message: 'Not connected to FireBot'
    });
  }

  try {
    await sendCommand(command);
    res.json({
      success: true,
      message: `Command ${command} sent`
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to send command: ' + error.message
    });
  }
});

/**
 * GET /api/status - Get current robot status
 */
app.get('/api/status', (req, res) => {
  res.json({
    connected: isConnected,
    sensorData: sensorData
  });
});

/**
 * GET /api/health - Health check endpoint
 */
app.get('/api/health', (req, res) => {
  res.json({
    status: 'running',
    server: 'FireBot Bluetooth Bridge',
    version: '1.0.0'
  });
});

// ========================================
// WebSocket Events
// ========================================

io.on('connection', (socket) => {
  console.log('🔌 WebSocket client connected:', socket.id);

  // Send current status to new client
  socket.emit('connection_status', { connected: isConnected });
  socket.emit('sensor_data', sensorData);

  // Handle command via WebSocket
  socket.on('send_command', async (data) => {
    try {
      await sendCommand(data.command);
      socket.emit('command_response', { success: true, command: data.command });
    } catch (error) {
      socket.emit('command_response', { success: false, command: data.command, error: error.message });
    }
  });

  socket.on('disconnect', () => {
    console.log('🔌 WebSocket client disconnected:', socket.id);
  });
});

// ========================================
// Server Startup
// ========================================

server.listen(PORT, () => {
  console.log('\n' + '='.repeat(50));
  console.log('🔥 FireBot Local Bluetooth Bridge Server (Node.js)');
  console.log('='.repeat(50));
  console.log(`Server running on: http://localhost:${PORT}`);
  console.log(`WebSocket available on: ws://localhost:${PORT}`);
  console.log('\nEndpoints:');
  console.log('  POST /api/connect       - Connect to ESP32');
  console.log('  POST /api/disconnect    - Disconnect from ESP32');
  console.log('  POST /api/send-command  - Send command to robot');
  console.log('  GET  /api/status        - Get robot status');
  console.log('  GET  /api/health        - Server health check');
  console.log('\nPress Ctrl+C to stop');
  console.log('='.repeat(50) + '\n');
  console.log('⚠️  IMPORTANT: Update BLUETOOTH_PORT in this file to match your ESP32 COM port');
  console.log('   Current: ' + BLUETOOTH_PORT);
  console.log('='.repeat(50) + '\n');
});

// Graceful shutdown
process.on('SIGINT', async () => {
  console.log('\n\n🛑 Shutting down server...');
  await disconnectBluetooth();
  process.exit(0);
});
